op-supervisor/supervisor/backend/db/sync/client.go

```diff
@@ -0,0 +1,226 @@
+type Client struct {
+	config  Config
+	baseURL string
+
+	httpClient *http.Client
+	retryDelay time.Duration
+}
+
+// NewClient creates a new Client with the given config and server URL.
+func NewClient(config Config, serverURL string, httpClient *http.Client) (*Client, error) {
+	// Verify root directory exists and is actually a directory
+	root, err := filepath.Abs(config.DataDir)
+	if err != nil {
+		return nil, fmt.Errorf("invalid root directory: %w", err)
+	}
+	rootInfo, err := os.Stat(root)
+	if err != nil {
+		return nil, fmt.Errorf("cannot access root directory: %w", err)
+	}
+	if !rootInfo.IsDir() {
+		return nil, fmt.Errorf("root path is not a directory: %s", root)
+	}
+
+	if httpClient == nil {
+		httpClient = DefaultHTTPClient
+	}
+
+	return &Client{
+		config:     config,
+		baseURL:    serverURL,
+		httpClient: httpClient,
+		retryDelay: InitRetryDelay,
+	}, nil
+}
+
+// SyncAll syncs all known databases for the given chains.
+func (c *Client) SyncAll(ctx context.Context, chains []types.ChainID, resume bool) error {
+	for _, chain := range chains {
+		for fileAlias := range Databases {
+			if err := c.SyncDatabase(ctx, chain, fileAlias, resume); err != nil {
+				return fmt.Errorf("failed to sync %s for chain %s: %w", fileAlias, chain, err)
+			}
+		}
+	}
+	return nil
+}
+
+// SyncDatabase downloads the named file from the server.
+// If the local file exists, it will attempt to resume the download if resume is true.
+func (c *Client) SyncDatabase(ctx context.Context, chainID types.ChainID, database Database, resume bool) error {
+	// Validate file alias
+	filePath, exists := Databases[database]
+	if !exists {
+		return fmt.Errorf("unknown file alias: %s", database)
+	}
+
+	// Ensure the chain directory exists
+	chainDir := filepath.Join(c.config.DataDir, chainID.String())
+	if err := os.MkdirAll(chainDir, 0755); err != nil {
+		return fmt.Errorf("failed to create chain directory: %w", err)
+	}
+
+	filePath = filepath.Join(chainDir, filePath)
+
+	// Ensure the database file exists and get initial size
+	filePath = filepath.Join(chainDir, filePath)
+	var initialSize int64
+	if stat, err := os.Stat(filePath); err == nil {
+		initialSize = stat.Size()
+	}
+
+	// If we have data already and don't want to resume then stop now
+	if initialSize > 0 && !resume {
+		return nil
+	}
+
+	// Keep track of the current retry delay
+	currentDelay := c.retryDelay
+	for {
+		// Try to sync and return if successful
+		err := c.attemptSync(ctx, chainID, database, filePath, initialSize)
+		if err == nil {
+			return nil
+		}
+
+		// Check if context was canceled
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		c.logError("sync attempt failed", err, database)
+
+		// Wait before retrying
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(currentDelay):
+			// Double the delay for next time, up to max
+			currentDelay *= 2
+			if currentDelay > MaxRetryDelay {
+				currentDelay = MaxRetryDelay
+			}
+		}
+	}
+}
```

Comment:
It seems like this loop will run forever since the supplied context is the backend context, and during creation I don't think anything is going to cancel it (other than stopping the program). I *think* that's ok? Because the other options don't seem like good UX either:
- Stop the Supervisor with an error about the sync endpoint
- Continue the Supervisor, syncing from scratch for that chain

Reply:
My original intention was to keep trying forever, with the philosophy being to always being trying to correct itself. However:

1) Using the monorepo's `retry` lib introduced a requirement to have some limit (currently set really high)
2) If the goal is to work toward making progress and we've failed for a long time, maybe we should just sync from the beginning. If it's too slow and the sync connection becomes available then they can restart.

I'm now thinking maybe we should try over the span of a few minutes and then move on to syncing from scratch. What do you think @axelKingsley @protolambda?

Reply:
I'm imagining an operator who sets the sync endpoint and walks away after launching the supervisor. When they come back and find the sync failed, my intuition is they would prefer for the Supervisor to exit with a clear message like "Sync was requested and needed, but the endpoint didn't respond after X attempts". At that point they can decide if they want to try again, fix it, or sync from genesis. If they sat down and found the application started syncing from genesis on its own, they'd have to go and *delete* the database file before they could try sync again. And if there's a deep chain history, they're unlikely to think of syncing from genesis as a reasonable option.

So my vote is for a high number of retries, followed by a clear exit error.